File: bindings/python/insight/signal/spectral.py
"""Spectral analysis functions for signal processing.

Provides power spectral density estimation (Welch, periodogram),
cross-spectral density, coherence, spectrogram, STFT, and vector strength.
"""

from insight._insight import signal as _signal

__all__ = [
    "csd",
    "welch",
    "periodogram",
    "coherence",
    "spectrogram",
    "stft",
    "vectorstrength",
]


def csd(
    x,
    y,
    fs=1.0,
    window="hann",
    nperseg=256,
    noverlap=None,
    nfft=None,
    detrend="constant",
    return_onesided=True,
    scaling="density",
):
    """Estimate the cross spectral density using Welch's method.

    Args:
        x: First input signal array.
        y: Second input signal array.
        fs: Sampling frequency in Hz.  Default is 1.0.
        window: Window function name.  Default is ``'hann'``.
        nperseg: Length of each segment.  Default is 256.
        noverlap: Number of overlapping points between segments.  If
            ``None`` (default), defaults to ``nperseg // 2``.
        nfft: FFT size.  If ``None`` (default), uses *nperseg*.
        detrend: Detrending method.  Default is ``'constant'``.
        return_onesided: If True (default), return a one-sided spectrum
            for real inputs.
        scaling: Scaling type.  ``'density'`` (default) for power spectral
            density; ``'spectrum'`` for power spectrum.

    Returns:
        A ``SpectralResult`` with attributes ``f`` (frequency array) and
        ``Pxx`` (cross spectral density).
    """
    kwargs = {}
    if noverlap is not None:
        kwargs["noverlap"] = noverlap
    if nfft is not None:
        kwargs["nfft"] = nfft
    return _signal.csd(
        x,
        y,
        fs=fs,
        window=window,
        nperseg=nperseg,
        detrend=detrend,
        return_onesided=return_onesided,
        scaling=scaling,
        **kwargs,
    )


def welch(
    x,
    fs=1.0,
    window="hann",
    nperseg=256,
    noverlap=None,
    nfft=None,
    detrend="constant",
    return_onesided=True,
    scaling="density",
):
    """Estimate the power spectral density using Welch's method.

    Args:
        x: Input signal array.
        fs: Sampling frequency in Hz.  Default is 1.0.
        window: Window function name.  Default is ``'hann'``.
        nperseg: Length of each segment.  Default is 256.
        noverlap: Number of overlapping points between segments.  If
            ``None`` (default), defaults to ``nperseg // 2``.
        nfft: FFT size.  If ``None`` (default), uses *nperseg*.
        detrend: Detrending method.  Default is ``'constant'``.
        return_onesided: If True (default), return a one-sided spectrum
            for real inputs.
        scaling: Scaling type.  ``'density'`` (default) for power spectral
            density; ``'spectrum'`` for power spectrum.

    Returns:
        A ``SpectralResult`` with attributes ``f`` (frequency array) and
        ``Pxx`` (power spectral density).
    """
    kwargs = {}
    if noverlap is not None:
        kwargs["noverlap"] = noverlap
    if nfft is not None:
        kwargs["nfft"] = nfft
    return _signal.welch(
        x,
        fs=fs,
        window=window,
        nperseg=nperseg,
        detrend=detrend,
        return_onesided=return_onesided,
        scaling=scaling,
        **kwargs,
    )


def periodogram(
    x, fs=1.0, window="hann", nfft=None, detrend="constant", return_onesided=True, scaling="density"
):
    """Estimate the power spectral density using a periodogram.

    Args:
        x: Input signal array.
        fs: Sampling frequency in Hz.  Default is 1.0.
        window: Window function name.  Default is ``'hann'``.
        nfft: FFT size.  If ``None`` (default), uses the signal length.
        detrend: Detrending method.  Default is ``'constant'``.
        return_onesided: If True (default), return a one-sided spectrum
            for real inputs.
        scaling: Scaling type.  ``'density'`` (default) for power spectral
            density; ``'spectrum'`` for power spectrum.

    Returns:
        A ``SpectralResult`` with attributes ``f`` (frequency array) and
        ``Pxx`` (power spectral density).
    """
    kwargs = {}
    if nfft is not None:
        kwargs["nfft"] = nfft
    return _signal.periodogram(
        x,
        fs=fs,
        window=window,
        detrend=detrend,
        return_onesided=return_onesided,
        scaling=scaling,
        **kwargs,
    )


def coherence(
    x, y, fs=1.0, window="hann", nperseg=256, noverlap=None, nfft=None, detrend="constant"
):
    """Estimate the magnitude-squared coherence between two signals.

    Args:
        x: First input signal array.
        y: Second input signal array.
        fs: Sampling frequency in Hz.  Default is 1.0.
        window: Window function name.  Default is ``'hann'``.
        nperseg: Length of each segment.  Default is 256.
        noverlap: Number of overlapping points between segments.  If
            ``None`` (default), defaults to ``nperseg // 2``.
        nfft: FFT size.  If ``None`` (default), uses *nperseg*.
        detrend: Detrending method.  Default is ``'constant'``.

    Returns:
        A ``SpectralResult`` with attributes ``f`` (frequency array) and
        ``Pxx`` (coherence values in [0, 1]).
    """
    kwargs = {}
    if noverlap is not None:
        kwargs["noverlap"] = noverlap
    if nfft is not None:
        kwargs["nfft"] = nfft
    return _signal.coherence(x, y, fs=fs, window=window, nperseg=nperseg, detrend=detrend, **kwargs)


def spectrogram(
    x,
    fs=1.0,
    window="hann",
    nperseg=256,
    noverlap=None,
    nfft=None,
    detrend="constant",
    return_onesided=True,
    mode="psd",
):
    """Compute a spectrogram using Welch's method.

    Args:
        x: Input signal array.
        fs: Sampling frequency in Hz.  Default is 1.0.
        window: Window function name.  Default is ``'hann'``.
        nperseg: Length of each segment.  Default is 256.
        noverlap: Number of overlapping points between segments.  If
            ``None`` (default), defaults to ``nperseg // 2``.
        nfft: FFT size.  If ``None`` (default), uses *nperseg*.
        detrend: Detrending method.  Default is ``'constant'``.
        return_onesided: If True (default), return a one-sided spectrum
            for real inputs.
        mode: Spectrogram mode.  One of ``'psd'`` (default), ``'magnitude'``,
            ``'angle'``, or ``'complex'``.

    Returns:
        A ``SpectrogramResult`` with attributes ``f`` (frequency array),
        ``t`` (time array), and ``Sxx`` (spectrogram array).
    """
    kwargs = {}
    if noverlap is not None:
        kwargs["noverlap"] = noverlap
    if nfft is not None:
        kwargs["nfft"] = nfft
    return _signal.spectrogram(
        x,
        fs=fs,
        window=window,
        nperseg=nperseg,
        detrend=detrend,
        return_onesided=return_onesided,
        mode=mode,
        **kwargs,
    )


def stft(x, fs=1.0, window="hann", nperseg=256, noverlap=None, nfft=None):
    """Compute the Short-Time Fourier Transform.

    Args:
        x: Input signal array.
        fs: Sampling frequency in Hz.  Default is 1.0.
        window: Window function name.  Default is ``'hann'``.
        nperseg: Length of each segment.  Default is 256.
        noverlap: Number of overlapping points between segments.  If
            ``None`` (default), defaults to ``nperseg // 2``.
        nfft: FFT size.  If ``None`` (default), uses *nperseg*.

    Returns:
        A ``SpectrogramResult`` with attributes ``f`` (frequency array),
        ``t`` (time array), and ``Sxx`` (STFT complex array).
    """
    kwargs = {}
    if noverlap is not None:
        kwargs["noverlap"] = noverlap
    if nfft is not None:
        kwargs["nfft"] = nfft
    return _signal.stft(x, fs=fs, window=window, nperseg=nperseg, **kwargs)


def vectorstrength(events, period):
    """Determine how well a set of events are synchronized to a period.

    Args:
        events: Array of event times.
        period: The period to which synchronization is measured.

    Returns:
        A tuple ``(strength, phase)`` where *strength* is the vector
        strength (0 = no synchronization, 1 = perfect) and *phase* is the
        mean phase in radians.
    """
    return _signal.vectorstrength(events, period)

File: bindings/lua/insight/signal/spectral.lua
--- Spectral analysis functions.
-- Provides Welch's method, periodogram, cross-spectral density,
-- coherence, spectrogram, STFT, and vector strength.
-- @module insight.signal.spectral

local native = require("_insight")
local sig = native.signal

local M = {}

local _wrap = require("insight._wrap")

--- Estimate power spectral density using Welch's method.
-- @tparam Array x Input signal array.
-- @number[opt=1.0] fs Sampling frequency.
-- @string[opt="hann"] window Window type.
-- @int[opt] nperseg Length of each segment.
-- @int[opt] noverlap Number of overlapping points.
-- @int[opt] nfft FFT length.
-- @string[opt="density"] scaling 'density' or 'spectrum'.
-- @treturn table Table with fields 'f' (frequencies) and 'Pxx' (PSD).
M.welch = _wrap(
  { "x", "fs", "window", "nperseg", "noverlap", "nfft", "scaling" },
  function(x, fs, window, nperseg, noverlap, nfft, scaling)
    return sig.welch(x, fs or 1.0, window or "hann", nperseg or 256, noverlap or 128, nfft or 256, scaling or "density")
  end
)

--- Estimate power spectral density using a periodogram.
-- @tparam Array x Input signal array.
-- @number[opt=1.0] fs Sampling frequency.
-- @string[opt="hann"] window Window type.
-- @int[opt] nfft FFT length.
-- @string[opt="density"] scaling 'density' or 'spectrum'.
-- @treturn table Table with fields 'f' (frequencies) and 'Pxx' (PSD).
M.periodogram = _wrap({ "x", "fs", "window", "nfft", "scaling" }, function(x, fs, window, nfft, scaling)
  return sig.periodogram(x, fs or 1.0, window or "hann", nfft or 256, scaling or "density")
end)

--- Estimate cross-spectral density using Welch's method.
-- @tparam Array x First input signal.
-- @tparam Array y Second input signal.
-- @number[opt=1.0] fs Sampling frequency.
-- @string[opt="hann"] window Window type.
-- @int[opt] nperseg Length of each segment.
-- @int[opt] noverlap Number of overlapping points.
-- @int[opt] nfft FFT length.
-- @treturn table Table with fields 'f' (frequencies) and 'Pxy' (CSD).
M.csd = _wrap(
  { "x", "y", "fs", "window", "nperseg", "noverlap", "nfft" },
  function(x, y, fs, window, nperseg, noverlap, nfft)
    return sig.csd(x, y, fs or 1.0, window or "hann", nperseg or 256, noverlap or 128, nfft or 256)
  end
)

--- Estimate magnitude-squared coherence using Welch's method.
-- @tparam Array x First input signal.
-- @tparam Array y Second input signal.
-- @number[opt=1.0] fs Sampling frequency.
-- @string[opt="hann"] window Window type.
-- @int[opt] nperseg Length of each segment.
-- @int[opt] noverlap Number of overlapping points.
-- @int[opt] nfft FFT length.
-- @treturn table Table with fields 'f' (frequencies) and 'Cxy' (coherence).
M.coherence = _wrap(
  { "x", "y", "fs", "window", "nperseg", "noverlap", "nfft" },
  function(x, y, fs, window, nperseg, noverlap, nfft)
    return sig.coherence(x, y, fs or 1.0, window or "hann", nperseg or 256, noverlap or 128, nfft or 256)
  end
)

--- Compute a spectrogram (SSTFT power spectral density).
-- @tparam Array x Input signal array.
-- @number[opt=1.0] fs Sampling frequency.
-- @string[opt="hann"] window Window type.
-- @int[opt] nperseg Length of each segment.
-- @int[opt] noverlap Number of overlapping points.
-- @int[opt] nfft FFT length.
-- @treturn table Table with fields 'f' (frequencies), 't' (times), 'Sxx' (spectrogram).
M.spectrogram = _wrap(
  { "x", "fs", "window", "nperseg", "noverlap", "nfft" },
  function(x, fs, window, nperseg, noverlap, nfft)
    return sig.spectrogram(x, fs or 1.0, window or "hann", nperseg or 256, noverlap or 128, nfft or 256)
  end
)

--- Compute the Short-Time Fourier Transform (STFT).
-- @tparam Array x Input signal array.
-- @number[opt=1.0] fs Sampling frequency.
-- @string[opt="hann"] window Window type.
-- @int[opt] nperseg Length of each segment.
-- @int[opt] noverlap Number of overlapping points.
-- @int[opt] nfft FFT length.
-- @treturn table Table with fields 'f' (frequencies), 't' (times), 'Zxx' (STFT).
M.stft = _wrap({ "x", "fs", "window", "nperseg", "noverlap", "nfft" }, function(x, fs, window, nperseg, noverlap, nfft)
  return sig.stft(x, fs or 1.0, window or "hann", nperseg or 256, noverlap or 128, nfft or 256)
end)

--- Compute the vector strength (phase coherence).
-- @tparam Array signal Input signal array.
-- @number freq Frequency in Hz.
-- @number fs Sampling frequency in Hz.
-- @treturn table Table with fields 'strength' and 'angle'.
M.vectorstrength = _wrap({ "signal", "freq", "fs" }, function(signal, freq, fs)
  return sig.vectorstrength(signal, freq, fs)
end)

return M
